src/finchlite/finch_assembly/dataflow.py

```diff
@@ -0,0 +1,150 @@
+            case Assert(exp):
+                exp_str = self._expr_str(exp, state)
+                return f"assert({exp_str})"
+            case _:
+                return str(stmt)
+
+    def _expr_str(self, expr, state: dict) -> str:
+        """Annotate an expression with lattice values."""
+        match expr:
+            case Literal(value):
+                return qual_str(value)
+            case TaggedVariable(Variable(name, _), id):
+                var_str = f"{name}_{id}"
+                if name in state:
+                    return f"{var_str}: {self.print_lattice_value(name, state)}"
+                return var_str
+            case Variable(name, _):
+                if name in state:
+                    return f"{name}: {self.print_lattice_value(name, state)}"
+                return str(name)
+            case Call(Literal(_) as lit, args):
+                annotated_args = [self._expr_str(arg, state) for arg in args]
+                return f"{qual_str(lit.val)}({', '.join(annotated_args)})"
+            case _:
+                return str(expr)
+
+    @abstractmethod
+    def print_lattice_value(self, name, state: dict) -> str:
+        """Format the lattice value associated with a variable for annotation."""
+        ...
+
+
+class AssemblyCopyPropagation(AbstractAssemblyDataflow):
+    def direction(self) -> str:
+        """
+        Copy propagation is a forward analysis.
+        """
+        return "forward"
+
+    def transfer(self, stmts, state: dict) -> dict:
+        new_state = state.copy()
+
+        for stmt in stmts:
+            match stmt:
+                case Assign(lhs, rhs):
+                    # Get the variable name being assigned to
+                    var_name = self._get_variable_name(lhs)
+
+                    if var_name is not None:
+                        new_state[var_name] = rhs
+
+                        # invalidate any copies that point
+                        # to the variable being assigned
+                        to_remove = []
+                        for var, val in new_state.items():
+                            if val == lhs:
+                                to_remove.append(var)
+
+                        for var in to_remove:
+                            new_state.pop(var)
+
+        return new_state
+
+    def join(self, state_1: dict, state_2: dict) -> dict:
+        result = {}
+
+        # only keep copy relationships that exist in both states with the same value
+        for var_name in state_1:
+            if var_name in state_2 and self._values_equal(
+                state_1[var_name], state_2[var_name]
+            ):
+                result[var_name] = state_1[var_name]
+
+        return result
+
+    def print_lattice_value(self, name, state: dict) -> str:
+        visited: set[str] = set()
+        value = state.get(name)
+
+        # iterate through states until we reach a Literal or a variable
+        # without a mapping (root value)
+        while True:
```

Comment:
This step indicates to me that the transfer function wasn't written correctly. You should always pass around the root variable for each copy, right?